aws/eks-anywhere
# Nutanix ControlPlaneEndpoint IP Validation

Add IP Uniqueness validation for ControlPlane Enpoint for Nutanix platform.

**How has this been tested?**
- Use a clusterspec with controlPlaneConfiguration.endpoint,host set to an IP that's already in use and run `eksctl-anywhere create cluster` and see it throw a validation failure.
```
$ bin/eksctl-anywhere create cluster -f eksa-sid.yaml
Performing setup and validations
odd number of arguments passed as key-value pairs for logging	{"ignored key": "eksa-sid"}
ValidateClusterSpec for Nutanix datacenter
❌ Validation failed	{"validation": "nutanix Provider setup is valid", "error": "cluster controlPlaneConfiguration.Endpoint.Host <10.40.142.14> is already in use, control plane IP must be unique", "errorVerbose": "cluster controlPlaneConfiguration.Endpoint.Host <10.40.142.14> is already in use, control plane IP must be unique\ngithub.com/aws/eks-anywhere/pkg/providers/validator.(*IPValidator).ValidateControlPlaneIPUniqueness\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/providers/validator/validate.go:42\ngithub.com/aws/eks-anywhere/pkg/providers/nutanix.(*Provider).SetupAndValidateCreateCluster\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/providers/nutanix/provider.go:203\ngithub.com/aws/eks-anywhere/pkg/workflows.(*SetAndValidateTask).providerValidation.func1\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/workflows/create.go:190\ngithub.com/aws/eks-anywhere/pkg/validations.(*Runner).Run\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/validations/runner.go:24\ngithub.com/aws/eks-anywhere/pkg/workflows.(*SetAndValidateTask).Run\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/workflows/create.go:177\ngithub.com/aws/eks-anywhere/pkg/task.(*taskRunner).RunTask\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/task/task.go:158\ngithub.com/aws/eks-anywhere/pkg/workflows.(*Create).Run\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/pkg/workflows/create.go:69\ngithub.com/aws/eks-anywhere/cmd/eksctl-anywhere/cmd.(*createClusterOptions).createCluster\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/cmd/eksctl-anywhere/cmd/createcluster.go:183\ngithub.com/spf13/cobra.(*Command).execute\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/vendor/github.com/spf13/cobra/command.go:916\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/vendor/github.com/spf13/cobra/command.go:1044\ngithub.com/spf13/cobra.(*Command).Execute\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/vendor/github.com/spf13/cobra/command.go:968\ngithub.com/spf13/cobra.(*Command).ExecuteContext\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/vendor/github.com/spf13/cobra/command.go:961\ngithub.com/aws/eks-anywhere/cmd/eksctl-anywhere/cmd.Execute\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/cmd/eksctl-anywhere/cmd/root.go:100\nmain.main\n\t/Users/sid.shukla/go/src/github.com/nutanix-cloud-native/aws-eks-anywhere/cmd/eksctl-anywhere/main.go:19\nruntime.main\n\t/usr/local/Cellar/go/1.19.3/libexec/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/Cellar/go/1.19.3/libexec/src/runtime/asm_amd64.s:1594", "remediation": ""}
✅ Validate OS is compatible with registry mirror configuration
✅ Validate certificate for registry mirror
✅ Validate authentication for git provider
Error: validations failed
```

- Use a clusterspec with controlPlaneConfiguration.endpoint,host set to an IP that's free and not in use and run `eksctl-anywhere create cluster` and see it pass the validation stage.
```
$ bin/eksctl-anywhere create cluster -f  eksa-sid.yaml
Performing setup and validations
odd number of arguments passed as key-value pairs for logging	{"ignored key": "eksa-sid"}
ValidateClusterSpec for Nutanix datacenter
✅ Nutanix Provider setup is valid
✅ Validate OS is compatible with registry mirror configuration
✅ Validate certificate for registry mirror
✅ Validate authentication for git provider
Creating new bootstrap cluster
Provider specific pre-capi-install-setup on bootstrap cluster
Installing cluster-api providers on bootstrap cluster
Provider specific post-setup
Creating new workload cluster
Installing networking on workload cluster
Creating EKS-A namespace
Installing cluster-api providers on workload cluster
Installing EKS-A secrets on workload cluster
Installing resources on management cluster
Moving cluster management from bootstrap to workload cluster
Installing EKS-A custom components (CRD and controller) on workload cluster
Installing EKS-D components on workload cluster
Creating EKS-A CRDs instances on workload cluster
Installing GitOps Toolkit on workload cluster
GitOps field not specified, bootstrap flux skipped
Writing cluster config file
Deleting bootstrap cluster
🎉 Cluster created!
--------------------------------------------------------------------------------------
The Amazon EKS Anywhere Curated Packages are only available to customers with the
Amazon EKS Anywhere Enterprise Subscription
--------------------------------------------------------------------------------------
Enabling curated packages on the cluster
Installing helm chart on cluster	{"chart": "eks-anywhere-packages", "version": "0.0.0-95a1c3cadbc2118660a5d08699d3c00c3680fe71-release-0.15-helm"}
```
